### PR TITLE
OZ-462: Ozone distro to repackage the O3 frontend

### PR DESCRIPTION
**Pull request by ibacher**
This is an attempt to fulfill OZ-462. I have not yet added docs, as there are some details I'm hoping to finalize.

**Assumptions:**

1. Machines running the distro have a recentish version Node / NPM (I think v16+) installed and available on the system PATH.
2. Ozone deployments do not actually need the customizations provided by spa-build-config.json.

**How this works:**

* Ozone now downloads the org.openmrs.distro:referenceapplication-frontend zip file and unpacks it in `$OZONE_DIR/configs/openmrs/frontend_assembly/reference-application-spa-assemble-config.json`
* An Ozone _distribution_ can provide a custom `spa-assemble-config.json` file in `$DISTRO_ROOT/configs/openmrs/frontend_assembly` that specifies their additions and removals.
* In an Ozone distribution, the `reference-application-spa-assemble-config.json` is copied to `${project.build.directory}/openmrs_frontend` and the `spa-assemble-config.json` file (if any) is copied to `${project.build.directory}/openmrs_frontend`.
* If a `spa-assemble-config.json` file is found, then the `assemble` (and, for now, at least, `build`) commands will be run. The outputs of this are published to `/distro/binaries/openmrs/frontend/`.

**Open questions:**

* Are these files in the right places or should they go somewhere else? (Little hard to know what to do with the assemble-configs; they aren't traditional config files in that they aren't part of the runtime, but they also aren't binaries for the same reason).
* Are the assumptions here ok?
* Do we need to always have some kind of output in `/distro/binaries/openmrs/frontend/`, e.g., so that this works with Docker?

(Apologies for adding yet more Groovy; I couldn't figure out how to do the necessary scripting in Ant)

### CODE REVIEW

**Comment by ibacher** [comment]
cc: @rbuisson @Ruhanga @enyachoke @corneliouzbett

**Comment by rbuisson** [comment]
>Do we need to always have some kind of output in /distro/binaries/openmrs/frontend/, e.g., so that this works with Docker?

Let's have the frontend always being rebuilt by Ozone even though based on default Ref App assemble JSON. That way we can always assume binaries will be present in the /distro/binaries/openmrs/frontend/ folder.

**Comment by rbuisson** [comment]
> Are the assumptions here ok?

As discussed in the call, we would want to be able to override the `SPA_CONFIG_URLS` and `DEFAULT_LOCALE`, but it looks like you've already pushed a fix for that. 👍🏻 .

**Comment by rbuisson** [comment]
>Are these files in the right places or should they go somewhere else? (Little hard to know what to do with the assemble-configs; they aren't traditional config files in that they aren't part of the runtime, but they also aren't binaries for the same reason).

I still hesitant on this one.
Let's keep it in `$OZONE_DIR/configs/openmrs/frontend_assembly/` and even keep it in the final package.
Even though it has no value at runtime, it's a very small file and won't hurt much.

**Comment by ibacher** [comment]
@rbuisson Another two questions (from the latest commit): I externalized the Groovy from the version embedded in the POM to an actual file, with the idea that it could be shared from Ozone to Ozone implementations and moved this to `scripts/openmrs/frontend_assembly` (though on reflection, it's probably better to go with `scripts/openmrs/frontend`).

1. Is that a reasonable place to put it?
2. Is it ok to publish scripts as part of the Ozone distro? (If so, I would probably use this same thing in the archetype to avoid the duplication of scripts)

**Comment by rbuisson** [comment]
@ibacher , 
> Is it ok to publish scripts as part of the Ozone distro? (If so, I would probably use this same thing in the archetype to avoid the duplication of scripts)

"Ozone distro" what package exactly? The final ozone? or ozone-distro?
OK to publish it but would it be possible to have it as a separate Maven package: "scripts".
What do you think?

**Comment by rbuisson** [comment]
>Is that a reasonable place to put it?

Works for me.

**Comment by ibacher** [comment]
> OK to publish it but would it be possible to have it as a separate Maven package: "scripts".

Oh, I like that idea a lot better. Let me refactor things again.

Yeah, I was meaning literally part of the ozone-distro artifact, but that's less than ideal.

**Comment by ibacher** [comment]
So I've added a `scripts` project to Ozone that bundles up `go-to-scripts-dir.sh`, `set-ozone-dir.sh` and the Groovy script to build / rebuild the frontend. Ozone (the distro) will use it as-is. Implementation projects rely on it being unpacked into the `target` directory.

File: scripts/pom.xml [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  <parent>
+    <groupId>com.ozonehis</groupId>
+    <artifactId>maven-commons</artifactId>
+    <version>1.0.0-SNAPSHOT</version>
+    <relativePath>../maven-commons</relativePath>
+  </parent>
+
+  <artifactId>ozone-scripts</artifactId>
+  <name>Ozone Scripts</name>
+  <description>Shared scripts for Ozone</description>
```
Comment (rbuisson) [inline scripts/pom.xml:16]: Could this be renamed to "Shared utility scripts for Ozone"?

File: ozone-distro-parent/pom.xml [hunk 2]
```diff
@@ -184,6 +226,25 @@
           </execution>
         </executions>
       </plugin>
+
+      <plugin>
+        <groupId>org.codehaus.gmavenplus</groupId>
+        <artifactId>gmavenplus-plugin</artifactId>
+        <executions>
+          <execution>
+            <id>Rebuild OpenMRS Frontend if necessary</id>
```
[the remark this answers is not in the document]
Comment (ibacher) [reply ozone-distro-parent/pom.xml]: This one only triggers if there are local customisations, but I can rename it.

File: maven-parent/pom.xml [hunk 3]
```diff
@@ -94,6 +112,49 @@
...
+              <goal>copy-resources</goal>
+            </goals>
+            <configuration>
+              <outputDirectory>
+                ${project.build.directory}/openmrs_frontend
+              </outputDirectory>
+              <overwrite>true</overwrite>
+              <resources>
+                <resource>
+                  <directory>${project.basedir}/configs/openmrs/frontend_assembly</directory>
```
[the remark this answers is not in the document]
Comment (ibacher) [reply maven-parent/pom.xml:148]: This step copies the implementation-specific `spa-assemble-config.json` file, if any. I didn't change that location, but I can if it makes more sense in `scripts`?

**Comment by rbuisson** [comment]
@ibacher , the frontend build seems to hang when running `./scripts/mvnw clean package`.

```
[INFO] Scanning for projects...
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Build Order:
[INFO] 
[INFO] Ozone Maven Commons                                                [pom]
[INFO] Ozone Distribution                                                 [pom]
[INFO] Ozone                                                              [pom]
[INFO] Ozone Scripts                                                      [pom]
[INFO] Ozone Maven Parent                                                 [pom]
[INFO] 
[INFO] ---------------------< com.ozonehis:maven-commons >---------------------
[INFO] Building Ozone Maven Commons 1.0.0-SNAPSHOT                        [1/5]
[INFO]   from maven-commons/pom.xml
[INFO] --------------------------------[ pom ]---------------------------------
[INFO] 
[INFO] --- clean:3.2.0:clean (default-clean) @ maven-commons ---
[INFO] 
[INFO] ---------------------< com.ozonehis:ozone-distro >----------------------
[INFO] Building Ozone Distribution 1.0.0-SNAPSHOT                         [2/5]
[INFO]   from distro/pom.xml
[INFO] --------------------------------[ pom ]---------------------------------
[INFO] 
[INFO] --- clean:3.2.0:clean (default-clean) @ ozone-distro ---
[INFO] Deleting /Users/rom/repos/ozone/distro/target
[INFO] 
[INFO] --- dependency:3.6.1:unpack-dependencies (Copy Odoo Initializer add-on) @ ozone-distro ---
[INFO] 
[INFO] --- dependency:3.6.1:copy-dependencies (Copy OpenMRS modules to a temporary location) @ ozone-distro ---
[INFO] Copying patientsummary-omod-2.2.0.jar to /Users/rom/repos/ozone/distro/target/openmrs_modules/patientsummary-omod-2.2.0.jar
[INFO] Copying commonreports-omod-1.4.0-SNAPSHOT.jar to /Users/rom/repos/ozone/distro/target/openmrs_modules/commonreports-omod-1.4.0-SNAPSHOT.jar
[INFO] 
[INFO] --- dependency:3.6.1:unpack-dependencies (Unpack OpenMRS Ref App in a temporary location) @ ozone-distro ---
[INFO] 
[INFO] --- dependency:3.6.1:unpack-dependencies (Unpack OpenMRS Ref App frontend configuration) @ ozone-distro ---
[INFO] 
[INFO] --- antrun:3.1.0:run (Rename spa-assemble-config.json to reference-application-spa-assemble-config.json) @ ozone-distro ---
[INFO] Executing tasks
[INFO]      [copy] Copying 1 file to /Users/rom/repos/ozone/distro/target/ozone-distro-1.0.0-SNAPSHOT/configs/openmrs/frontend_assembly
[INFO]      [move] Moving 1 file to /Users/rom/repos/ozone/distro/target/openmrs_frontend
[INFO] Executed tasks
[INFO] 
[INFO] --- resources:3.3.1:copy-resources (Exclude some OpenMRS Ref App files) @ ozone-distro ---
[INFO] Copying 71 resources from target/referenceapplication-distro/openmrs_config to target/openmrs_config
[INFO] 
[INFO] --- antrun:3.1.0:run (Rename JAR to OMOD) @ ozone-distro ---
[INFO] Executing tasks
[INFO]      [move] Moving 2 files to /Users/rom/repos/ozone/distro/target/openmrs_modules
[INFO] Executed tasks
[INFO] 
[INFO] --- gplus:3.0.2:execute (Build OpenMRS Frontend) @ ozone-distro ---
[INFO] Using plugin classloader, includes GMavenPlus and project classpath.
[INFO] Using Groovy 4.0.18 to perform execute.
[INFO] Running Groovy script from file:/Users/rom/repos/ozone/distro/../scripts/openmrs/frontend_assembly/build-openmrs-frontend.groovy.
[INFO] Checking for frontend customizations in /Users/rom/repos/ozone/distro/target/openmrs_frontend/spa-assemble-config.json
[INFO] Project: com.ozonehis:ozone-distro
[INFO] Running assemble command...
npx: installed 695 in 12.414s
[09:27:16] using cwd ~/repos/ozone
[09:27:16] updated package.json config with {}
[09:27:16] starting assemble 
[09:27:16] starting assemble:default task
[09:27:16] starting assemble:prompt-new task
```
and it hangs here.

**Comment by ibacher** [comment]
@rbuisson What version of Node do you have installed? (See the assumptions in the initial post of this PR).

**Comment by rbuisson** [comment]
Yep. What using a 17.x. Works fine with 18.x.

File: maven-parent/pom.xml (hunk 3 above)
[the remark this answers is not in the document]
Comment (rbuisson) [reply maven-parent/pom.xml:148]: That's OK we can leave it at this place.
But shouldn't this be `distro/configs/openmrs/frontend_assembly`.

**Comment by ibacher** [comment]
> Yep. What using a 17.x. Works fine with 18.x.

Ah... I should revise the comment. We support the LTS versions of Node, which honestly means "the even version numbers" (Node's versioning policy is that every other major release is LTS). I tested it with 16 and 18, but not 17.

File: maven-parent/pom.xml (hunk 3 above)
[the remark this answers is not in the document]
Comment (ibacher) [reply maven-parent/pom.xml:148]: In the implementation-specific projects, we usually don't have a `distro` sub-folder...

**Pull request merged**